File: leetcode/math/2613_Beautiful_Pairs/codes/seg_tree.c
#include <stdlib.h>
#include <limits.h>

typedef struct {
    int val;
    int idx;
} Pair;

int cmpPair(const void *a, const void *b) {
    const Pair *pa = (const Pair*)a;
    const Pair *pb = (const Pair*)b;
    if (pa->val != pb->val) return pa->val - pb->val;
    return pa->idx - pb->idx;
}

int better(const long *v, int x, int y) {
    if (x < 0) return y;
    if (y < 0) return x;
    if (v[x] < v[y] || (v[x] == v[y] && y < x)) return y;
    return x;
}

int query_tree(int *tree, const long *v, int ind, int l, int r, int ql, int qr) {
    if (ql > qr) return -1;
    if (l >= ql && r <= qr) return tree[ind];
    int mid = (l + r) >> 1;
    int res = -1;
    if (ql <= mid) res = query_tree(tree, v, ind<<1, l, mid, ql, qr);
    if (qr > mid) res = better(v, res, query_tree(tree, v, (ind<<1)|1, mid+1, r, ql, qr));
    return res;
}

void update_tree(int *tree, const long *v, int ind, int l, int r, int pos, int val) {
    tree[ind] = better(v, tree[ind], val);
    if (l == r) return;
    int mid = (l + r) >> 1;
    if (pos <= mid) update_tree(tree, v, ind<<1, l, mid, pos, val);
    else update_tree(tree, v, (ind<<1)|1, mid+1, r, pos, val);
}

int* beautifulPair(int* nums1, int nums1Size, int* nums2, int nums2Size, int* returnSize) {
    int n = nums1Size;
    long *p1 = malloc(sizeof(long)*n);
    long *p2 = malloc(sizeof(long)*n);
    int i;
    int maxCoord = 0;
    for (i = 0; i < n; ++i) {
        p1[i] = (long)nums2[i] + (long)nums1[i];
        p2[i] = (long)nums2[i] - (long)nums1[i];
        if (nums1[i] > maxCoord) maxCoord = nums1[i];
    }
    int size = maxCoord + 2;
    int treeSize = 4*(size+5);
    int *tree1 = malloc(sizeof(int)*treeSize);
    int *tree2 = malloc(sizeof(int)*treeSize);
    for (i = 0; i < treeSize; ++i) { tree1[i] = -1; tree2[i] = -1; }

    Pair *arr = malloc(sizeof(Pair)*n);
    for (i = 0; i < n; ++i) {
        arr[i].val = nums2[i];
        arr[i].idx = i;
    }
    qsort(arr, n, sizeof(Pair), cmpPair);

    int ans0 = n, ans1 = n;
    long long d = LLONG_MAX;
    for (i = 0; i < n; ++i) {
        int id = arr[i].idx;
        int x = nums1[id];
        if (x >= 0) {
            int i1 = query_tree(tree1, p1, 1, 0, maxCoord+1, 0, x);
            if (i1 >= 0) {
                long long may = p1[id] - p1[i1];
                if (may < d || (may == d && ( ( (id<i1?id:i1) < ans0) || (((id<i1?id:i1) == ans0) && ((id>i1?id:i1) < ans1)) ))) {
                    d = may; ans0 = id<i1?id:i1; ans1 = id>i1?id:i1;
                }
            }
        }
        int i2 = query_tree(tree2, p2, 1, 0, maxCoord+1, x, maxCoord+1);
        if (i2 >= 0) {
            long long may = p2[id] - p2[i2];
            if (may < d || (may == d && ( ( (id<i2?id:i2) < ans0) || (((id<i2?id:i2) == ans0) && ((id>i2?id:i2) < ans1)) ))) {
                d = may; ans0 = id<i2?id:i2; ans1 = id>i2?id:i2;
            }
        }
        update_tree(tree1, p1, 1, 0, maxCoord+1, x, id);
        update_tree(tree2, p2, 1, 0, maxCoord+1, x, id);
    }
    free(p1); free(p2); free(tree1); free(tree2); free(arr);

    int *res = malloc(sizeof(int)*2);
    res[0] = ans0; res[1] = ans1;
    *returnSize = 2;
    return res;
}

File: leetcode/math/2613_Beautiful_Pairs/codes/seg_tree.py
from typing import List

class Solution:
    def beautifulPair(self, nums1: List[int], nums2: List[int]) -> List[int]:
        n = len(nums1)
        p1 = [nums2[i] + nums1[i] for i in range(n)]
        p2 = [nums2[i] - nums1[i] for i in range(n)]
        maxCoord = max(nums1) if n>0 else 0
        size = maxCoord + 1
        tree1 = [-1] * (4 * (size+1))
        tree2 = [-1] * (4 * (size+1))
        def better(v, x, y):
            if x < 0:
                return y
            if y < 0:
                return x
            if v[x] < v[y] or (v[x] == v[y] and y < x):
                return y
            return x
        def query(tree, v, ind, left, right, ql, qr):
            if ql > qr:
                return -1
            if left >= ql and right <= qr:
                return tree[ind]
            mid = (left + right) >> 1
            res = -1
            if ql <= mid:
                res = query(tree, v, ind<<1, left, mid, ql, qr)
            if qr > mid:
                res = better(v, res, query(tree, v, (ind<<1)|1, mid+1, right, ql, qr))
            return res
        def update(tree, v, ind, left, right, pos, val):
            tree[ind] = better(v, tree[ind], val)
            if left == right:
                return
            mid = (left + right) >> 1
            if pos <= mid:
                update(tree, v, ind<<1, left, mid, pos, val)
            else:
                update(tree, v, (ind<<1)|1, mid+1, right, pos, val)
        idx = list(range(n))
        idx.sort(key=lambda x: nums2[x])
        ans_i1 = n
        ans_i2 = n
        d = 2**63-1
        for k in range(n):
            i = idx[k]
            x = nums1[i]
            if x >= 0:
                i1 = query(tree1, p1, 1, 0, size, 0, x)
                if i1 >= 0:
                    may = p1[i] - p1[i1]
                    if may < d or (may == d and (min(i, i1) < ans_i1 or (min(i, i1) == ans_i1 and max(i, i1) < ans_i2))):
                        d = may
                        ans_i1, ans_i2 = min(i, i1), max(i, i1)
            i2 = query(tree2, p2, 1, 0, size, x, size)
            if i2 >= 0:
                may = p2[i] - p2[i2]
                if may < d or (may == d and (min(i, i2) < ans_i1 or (min(i, i2) == ans_i1 and max(i, i2) < ans_i2))):
                    d = may
                    ans_i1, ans_i2 = min(i, i2), max(i, i2)
            update(tree1, p1, 1, 0, size, x, i)
            update(tree2, p2, 1, 0, size, x, i)
        return [ans_i1, ans_i2]
